RecoLocalCalo/HcalRecAlgos/interface/MahiFit.h

```diff
@@ -0,0 +1,167 @@
+
+  //holds active bunch crossings
+  BXVector bxs;  
+
+  //holds data samples
+  SampleVector amplitudes;
+
+  //holds inverse covariance matrix
+  SampleMatrix invCovMat;
+
+  //holds diagonal noise terms
+  SampleVector noiseTerms;
+
+  //holds flat pedestal uncertainty
+  SampleMatrix pedConstraint;
+  
+  //holds full covariance matrix for a pulse shape 
+  //varied in time
+  std::array<FullSampleMatrix, MaxPVSize> pulseCovArray;
+
+  //holds full pulse shape template
+  std::array<FullSampleVector, MaxPVSize> pulseShapeArray;
+
+  //holds full pulse shape derivatives
+  std::array<FullSampleVector, MaxPVSize> pulseDerivArray;
+
+  //holders for calculating pulse shape & covariance matrices
+  std::array<double, MaxSVSize> pulseN;
+  std::array<double, MaxSVSize> pulseM;
+  std::array<double, MaxSVSize> pulseP;
+
+  //holds matrix of pulse shape templates for each BX
+  SamplePulseMatrix pulseMat;
+
+  //holds matrix of pulse shape derivatives for each BX
+  SamplePulseMatrix pulseDerivMat;
+
+  //holds residual vector
+  PulseVector residuals;
+
+  //for FNNLS algorithm
+  unsigned int nP;
+  PulseVector ampVec;
+
+  PulseVector errVec;
+  PulseVector ampvecpermtest;
+
+  SamplePulseMatrix invcovp;
+  PulseMatrix aTaMat; // A-transpose A (matrix)
+  PulseVector aTbVec; // A-transpose b (vector)
+  PulseVector updateWork; // w (vector)
+
+  SampleDecompLLT covDecomp;
+  SampleMatrix covDecompLinv;
+  PulseMatrix topleft_work;
+  PulseDecompLDLT pulseDecomp;
+
+};
+
+class MahiFit
+{
+ public:
+  MahiFit();
+  ~MahiFit() { };
+
+  void setParameters(bool iDynamicPed, double iTS4Thresh, double chiSqSwitch, 
+		     bool iApplyTimeSlew, HcalTimeSlew::BiasSetting slewFlavor,
+		     double iMeanTime, double iTimeSigmaHPD, double iTimeSigmaSiPM, 
+		     const std::vector <int> &iActiveBXs, int iNMaxItersMin, int iNMaxItersNNLS,
+		     double iDeltaChiSqThresh, double iNnlsThresh);
+
+  void phase1Apply(const HBHEChannelInfo& channelData, 
+		   float& reconstructedEnergy, 
+		   float& reconstructedTime, 
+		   bool& useTriple,
+		   float& chi2) const;
+
+  void doFit(std::vector<float> &correctedOutput, int nbx) const;
+
+  void setPulseShapeTemplate  (const HcalPulseShapes::Shape& ps);
+  void resetPulseShapeTemplate(const HcalPulseShapes::Shape& ps);
+
+  typedef BXVector::Index Index;
+  const HcalPulseShapes::Shape* currentPulseShape_=nullptr;
+
+ private:
+
+  double minimize() const;
+  void onePulseMinimize() const;
+  void updateCov() const;
+  void updatePulseShape(double itQ, FullSampleVector &pulseShape, 
+			FullSampleVector &pulseDeriv,
+			FullSampleMatrix &pulseCov) const;
+  double calculateArrivalTime() const;
+  double calculateChiSq() const;
+  void nnls() const;
+  void resetWorkspace() const;
+
+  void nnlsUnconstrainParameter(Index idxp) const;
+  void nnlsConstrainParameter(Index minratioidx) const;
+
+  void eigenSolveSubmatrix(PulseMatrix& mat, PulseVector& invec, PulseVector& outvec, unsigned NP) const;
```

Comment:
probably best not to include eigen in the name of this function incase it changes someday